Enable TestDefaultBinaryMessenger to intercept all platform channels.

HelioStrike opened the pull request:
This PR defines `AllMessageHandler`, and adds a method `setMockAllMessagesHandler` to `TestDefaultBinaryMessenger`. This method allows a test to intercept all platform channel calls made during the duration of the test.

Solves #94853.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

I could not find where to add documentation (outside code comments), if I should. 

I also couldn't find a test testing `setMockMethodCallHandler`, and `setMockAllMessagesHandler` is sort of an extension of that, so I wasn't sure how to test this. Any suggestions are welcome!

Thanks!

Commit:
Enable TestDefaultBinaryMessenger to intercept all platform channels.

HelioStrike commented:
Figured how to test soon after I created the PR :'D

goderbauer commented:
@HelioStrike Thanks for the contribution. Can you please take a look at the failing check and fix what it is complaining about? Thanks!

HelioStrike commented:
@goderbauer Fixed analyzer issues by using "flutter analyze" locally. Hoping that solves it.

HelioStrike commented:
analyze-linux seems to be failing with what appears to be some sort of initialisation error, is there any way I can restart it? Same with luci-flutter. 

If the failures are because of issues with the PR, do you folks have any suggestions for debugging what's wrong?

goderbauer commented:
I have restarted the check. If that doesn't do it, you may have to rebase this PR to the latest master.

goderbauer commented:
I also noticed that this PR contains a lot of unrelated and undesired format changes. Can you please undo those to make reviewing easier?

HelioStrike commented:
Sorry for the delayed update (something wrong with my gmail notifications), but I removed the formatting changes and rebased on top of master. Thanks!

HelioStrike commented:
@goderbauer The checks are passing now :D

goderbauer commented on packages/flutter_test/lib/src/test_default_binary_messenger.dart:
```diff
@@ -8,6 +8,11 @@ import 'dart:ui' as ui;
 
+/// A function which takes the name of the method channel, it's handler,
+/// platform message and asynchronously returns an encoded response.
+typedef AllMessagesHandler = Future<ByteData?>? Function(
+    String channel, MessageHandler? handler, ByteData? message);
```
formatting nit:

```suggestion
typedef AllMessagesHandler = Future<ByteData?>? Function(
  String channel, 
  MessageHandler? handler,
  ByteData? message,
);
```

goderbauer commented on packages/flutter_test/lib/src/test_default_binary_messenger.dart:
```diff
@@ -116,11 +121,17 @@ class TestDefaultBinaryMessenger extends BinaryMessenger {
   // can implement the [checkMockMessageHandler] method.
   final Map<String, Object> _outboundHandlerIdentities = <String, Object>{};
 
+  /// Handler that intercepts and responds to outgoing messages, pretending
```
This doc needs a little more details describing how to use this and how setting this affects how the regular handler gets called.

goderbauer replied:
There should also be a note on setMockMessageHandler linking back to this.

And this doc should probably link to setMockMessageHandler as well.

goderbauer replied:
Some of the information included at https://master-api.flutter.dev/flutter/flutter_test/TestDefaultBinaryMessenger/setMockMessageHandler.html also belongs here, I believe.

HelioStrike replied:
Thanks for the review! :)
Added some more documentation, WDYT?

goderbauer commented on packages/flutter_test/lib/src/test_default_binary_messenger.dart:
```diff
@@ -116,11 +121,17 @@ class TestDefaultBinaryMessenger extends BinaryMessenger {
 
+  /// Handler that intercepts and responds to outgoing messages, pretending
+  /// to be the platform, for all channels.
+  AllMessagesHandler? allMessagesHandler;
```
setMockMessageHandler documents that "Registered callbacks are cleared after each test.". This one should probably as well, if it isn't already.

HelioStrike replied:
I couldn't find any references of setMockMessageHandler (or references of references which) in the flutter repo that did that. Any suggestions on where this might be happening?

goderbauer commented on packages/flutter_test/lib/src/test_default_binary_messenger.dart:
```diff
@@ -8,6 +8,11 @@ import 'dart:ui' as ui;
 import 'package:flutter/services.dart';
 
+/// A function which takes the name of the method channel, it's handler,
+/// platform message and asynchronously returns an encoded response.
```
document in what cases handler could be null.

goderbauer replied:
Can you please address this comment?

HelioStrike replied:
Ah, sorry, I thought this comment was on the allMessagesHandler and not the type, and thought it was addressed. Added now :)

goderbauer reviewed (commented):
@dnfield Curious to hear what you think about adding this to the test framework.

In theory, you could achieve the same effect by providing a custom binding that just overrides the createBinaryMessanger method.

dnfield commented:
This seems like it's probably fine. In some environments it's more difficult to override the default binding.

HelioStrike commented:
@goderbauer Fixed  analyzer issues, and all checks seem to be passing now too.

goderbauer reviewed (approved):
LGTM after the last doc comment is addressed.

goderbauer commented:
I restarted the google testing check, hopefully it will complete this time. Sorry for the delay.

goderbauer commented:
@HelioStrike Could you rebase this change with the latest master? Somehow one of the checks got stuck and that should trigger the check again. Thanks!

HelioStrike commented:
@goderbauer Hey! I'm on vacation this week, and am away from my laptop (commenting using my phone). I'll update by Monday

HelioStrike commented:
@goderbauer Sorry about the delay. Rebased now!

goderbauer commented:
@dnfield Wanna give this one a secondary review?
